src/solvers/refinement/string_constraint_generator_code_points.cpp

```diff
@@ -0,0 +1,239 @@
+
+#include <solvers/refinement/string_constraint_generator.h>
+
+/*******************************************************************    \
+
+Function: string_constraint_generatort::add_axioms_for_code_point
+
+  Inputs: an expression representing a java code point
+
+ Outputs: a new string expression
+
+ Purpose: add axioms for the conversion of an integer representing a java
+          code point to a utf-16 string
+
+\*******************************************************************/
+
+string_exprt string_constraint_generatort::add_axioms_for_code_point(
+  const exprt &code_point)
+{
+  string_exprt res(get_char_type());
+  const typet &type=code_point.type();
+  assert(type.id()==ID_signedbv);
+
+  // We add axioms:
+  // a1 : code_point<0x010000 => |res|=1
+  // a2 : code_point>=0x010000 => |res|=2
+  // a3 : code_point<0x010000 => res[0]=code_point
+  // a4 : code_point>=0x010000 => res[0]=0xD800+(code_point-0x10000)/0x0400
+  // a5 : code_point>=0x010000 => res[1]=0xDC00+(code_point-0x10000)/0x0400
+  // For more explenations about this conversion, see:
+  //   https://en.wikipedia.org/wiki/UTF-16
+
+  const exprt &hex010000=from_integer(0x010000, type);
+  const exprt &hexD800=from_integer(0xD800, type);
+  const exprt &hexDC00=from_integer(0xDC00, type);
+  const exprt &hex0400=from_integer(0x0400, type);
+
+  binary_relation_exprt small(code_point, ID_lt, hex010000);
+  implies_exprt a1(small, res.axiom_for_has_length(1));
+  axioms.push_back(a1);
+
+  implies_exprt a2(not_exprt(small), res.axiom_for_has_length(2));
+  axioms.push_back(a2);
+
+  typecast_exprt code_point_as_char(code_point, get_char_type());
+  implies_exprt a3(small, equal_exprt(res[0], code_point_as_char));
+  axioms.push_back(a3);
+
+  plus_exprt first_char(
+    hexD800, div_exprt(minus_exprt(code_point, hex010000), hex0400));
+  implies_exprt a4(
+    not_exprt(small),
+    equal_exprt(res[0], typecast_exprt(first_char, get_char_type())));
+  axioms.push_back(a4);
+
+  plus_exprt second_char(hexDC00, mod_exprt(code_point, hex0400));
+  implies_exprt a5(
+    not_exprt(small),
+    equal_exprt(res[1], typecast_exprt(second_char, get_char_type())));
+  axioms.push_back(a5);
+
+  return res;
+}
+
+
+exprt string_constraint_generatort::is_high_surrogate(const exprt &chr) const
+{
+  return and_exprt(
+    binary_relation_exprt(chr, ID_ge, constant_char(0xD800)),
+    binary_relation_exprt(chr, ID_le, constant_char(0xDBFF)));
+}
+
+exprt string_constraint_generatort::is_low_surrogate(const exprt &chr) const
+{
+  return and_exprt(
+    binary_relation_exprt(chr, ID_ge, constant_char(0xDC00)),
+    binary_relation_exprt(chr, ID_le, constant_char(0xDFFF)));
+}
+
+// pair_value=0x10000+(((char1%0x0800)*0x0400)+char2%0x0400)
+// char1 and char2 should be of type return_type
+exprt pair_value(exprt char1, exprt char2, typet return_type)
+{
+  const exprt &hex010000=from_integer(0x010000, return_type);
```

Comment:
I removed these kind of assignments (ie declaring `exprt hex010000` instead of `const exprt &hex010000`) for new values that are returned by functions https://github.com/diffblue/cbmc/pull/374/commits/fbb2108521ac9f1f0c16aaa0332173147609dbd6